OIDC CodeFlowTest has become unstable

Hi All, so I've continued tracing and I think I've reached the point where I have no idea what else can be done to trace it. Here is what is happening when a successful `testAccessTokenInjectionTest` is run:
```
// Get a static resource index.html, OIDC CodeAuthenticationMechanism checks it
(1)    2020-05-20 13:27:19,281 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-7) Auth:http://localhost:8081/index.html

// PathMatchingHttpSecurityPolicy checks (2) and denies (3) an acess to index.html
(2) 2020-05-20 13:27:19,291 ERROR [io.qua.ver.htt.run.sec.PathMatchingHttpSecurityPolicy] (vert.x-eventloop-thread-7) p1:http://localhost:8081/index.html
(3) 2020-05-20 13:27:19,294 ERROR [io.qua.ver.htt.run.sec.PathMatchingHttpSecurityPolicy] (vert.x-eventloop-thread-7) p2:http://localhost:8081/index.html

// This is the actual test, here we are already at the Keycloak Login page, because invisibly to the user  CodeAuthenticationMechanism has redirect the user to authenticate 
(4) 2020-05-20 13:27:20,180 ERROR [io.qua.it.key.CodeFlowTest] (main) Login***

// This is a user returning back from Keycloak, part of the code flow, note URI is http://localhost:8081/web-app per the code flow configuration, but since the user originally requested index.html, then, by default, we restore that path, see the next step

(5) 2020-05-20 13:27:20,561 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-7) Auth:http://localhost:8081/web-app?state=3042c883-58d6-40cf-9d3d-cc93dece5e23___%2Findex.html&session_state=02582d1c-c40d-4a2c-b5f1-b7616e449c34&code=afe72ae1-36f8-4b24-bcc4-5087e48726cf.02582d1c-c40d-4a2c-b5f1-b7616e449c34.693b9131-8da6-4d72-a81c-43ef231582e8

// This is CodeAuthenticationMechanism again, note now we are accessing index.html, as in step 1
 
(6) 2020-05-20 13:27:20,566 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-7) Auth:http://localhost:8081/index.html?pathChecked=true&state=3042c883-58d6-40cf-9d3d-cc93dece5e23___%2Findex.html&session_state=02582d1c-c40d-4a2c-b5f1-b7616e449c34&code=afe72ae1-36f8-4b24-bcc4-5087e48726cf.02582d1c-c40d-4a2c-b5f1-b7616e449c34.693b9131-8da6-4d72-a81c-43ef231582e8

// This is where we have redirected the user again to drop the redundant query parameters - at this stage the user already has a session authentication cookie, so CodeAuthenticationMechanism just lets it through 
(7) 2020-05-20 13:27:20,618 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-7) Auth:http://localhost:8081/index.html

// PathMatchingHttpSecurityPolicy checks it, all is good
(8) 2020-05-20 13:27:20,624 ERROR [io.qua.ver.htt.run.sec.PathMatchingHttpSecurityPolicy] (vert.x-eventloop-thread-7) p1:http://localhost:8081/index.html

// the test continues, after `index.html` has been successfully retrieved, we check the JAX-RS endpoint
// since the session auth cookie is still there,  CodeAuthenticationMechanism is happy
(9) 2020-05-20 13:27:20,660 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-7) Auth:http://localhost:8081/web-app/access

// PathMatchingHttpSecurityPolicy is also happy
(10) 2020-05-20 13:27:20,662 ERROR [io.qua.ver.htt.run.sec.PathMatchingHttpSecurityPolicy] (vert.x-eventloop-thread-7) p1:http://localhost:8081/web-app/access

(11) End of the test
```

So I run it in a long loop, and I have say 450 of the above blocks logged in exactly the same sequence.
And then, out of the blue, the following is reported:
```
[ERROR] testAccessTokenInjection[219]  Time elapsed: 0.51 s  <<< ERROR!
com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException: 401 Unauthorized for http://localhost:8081/index.html
	at com.gargoylesoftware.htmlunit.WebClient.throwFailingHttpStatusCodeExceptionIfNecessary(WebClient.java:595)
	at com.gargoylesoftware.htmlunit.WebClient.loadDownloadedResponses(WebClient.java:2277)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.doProcessPostponedActions(JavaScriptEngine.java:914)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.processPostponedActions(JavaScriptEngine.java:1002)
	at com.gargoylesoftware.htmlunit.html.DomElement.click(DomElement.java:1047)
	at com.gargoylesoftware.htmlunit.html.DomElement.click(DomElement.java:989)
	at com.gargoylesoftware.htmlunit.html.DomElement.click(DomElement.java:897)
	at com.gargoylesoftware.htmlunit.html.DomElement.click(DomElement.java:878)
	at com.gargoylesoftware.htmlunit.html.DomElement.click(DomElement.java:859)
	at io.quarkus.it.keycloak.CodeFlowTest.testAccessTokenInjection(CodeFlowTest.java:452)
```
with none of my error log messages displayed. For example, I'd N+ of those 11 steps logged, and then boom, it looks like as if some stale 401 came back. I can't trace where this error came from. neither `PathMatchingHttpSecurityPolicy` nor `CodeAuthenticationMechanism` is apparently called.

@stuartwdouglas do you have some hints ?

Here is a tracing sequence at the CORS filter level with the successful test following by 401:
```
(1) Get index.html, get 302 redirect to KC
 2020-05-20 15:38:49,934 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) Begin: http://localhost:8081/index.html
2020-05-20 15:38:49,934 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-12) Auth: http://localhost:8081/index.html
2020-05-20 15:38:49,934 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) End: http://localhost:8081/index.html 302

(2) User has authenticated, and returning back from KC, it will end with 302 to restore the original path

2020-05-20 15:38:50,439 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) Begin: http://localhost:8081/web-app?state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e
2020-05-20 15:38:50,439 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-12) Auth:http://localhost:8081/web-app?state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e
2020-05-20 15:38:50,439 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) End: http://localhost:8081/web-app?state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e 302

(3) index.html restored, another 302 to drop the query parameters

2020-05-20 15:38:50,439 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) Begin: http://localhost:8081/index.html?pathChecked=true&state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e
2020-05-20 15:38:50,439 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-12) Auth:http://localhost:8081/index.html?pathChecked=true&state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e
2020-05-20 15:38:50,446 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) End: http://localhost:8081/index.html?pathChecked=true&state=8a40a335-568b-4d57-9fab-48ab468deab7___%2Findex.html&session_state=4e98e01e-09ee-4815-af0a-44930fb6e4fe&code=9e3e11f0-5b97-4786-8797-83067fdc7e15.4e98e01e-09ee-4815-af0a-44930fb6e4fe.c61d7508-b25b-401e-bcc0-e1cd226c0b9e 302

(4) Finally we are getting index.html with 200

2020-05-20 15:38:50,447 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) Begin: http://localhost:8081/index.html
2020-05-20 15:38:50,447 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-12) Auth:http://localhost:8081/index.html
2020-05-20 15:38:50,448 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) End: http://localhost:8081/index.html 200

(5) Test continues, checking JAX-RS endpoint
2020-05-20 15:38:50,453 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (vert.x-eventloop-thread-12) Begin: http://localhost:8081/web-app/access
2020-05-20 15:38:50,453 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-12) Auth:http://localhost:8081/web-app/access
2020-05-20 15:38:50,454 ERROR [io.qua.ver.htt.run.cor.CORSFilter] (executor-thread-1) End: http://localhost:8081/web-app/access 200

(6) all is good, and now:
[ERROR] Tests run: 515, Failures: 0, Errors: 1, Skipped: 14, Time elapsed: 255.705 s <<< FAILURE! - in io.quarkus.it.keycloak.CodeFlowInGraalITCase
[ERROR] testAccessTokenInjection[259]  Time elapsed: 0.487 s  <<< ERROR!
com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException: 401 Unauthorized for http://localhost:8081/index.html

Where it came from ? This is something with the Vertx ending the request, may be not

```

This is weird now:

```
unnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.security.SignatureException: Signature length not correct: got 25 but was expecting 256
        at sun.security.rsa.RSASignature.engineVerify(RSASignature.java:211)
        at java.security.Signature$Delegate.engineVerify(Signature.java:1394)
        at java.security.Signature.verify(Signature.java:771)
        at io.vertx.ext.jwt.JWK.verify(JWK.java:412)
        ... 85 more

```

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -179,7 +179,7 @@
         <quarkus-spring-security-api.version>5.2.Final</quarkus-spring-security-api.version>
         <quarkus-spring-boot-api.version>2.1.SP1</quarkus-spring-boot-api.version>
         <mvel2.version>2.4.7.Final</mvel2.version>
-        <htmlunit.version>2.36.0</htmlunit.version>
+        <htmlunit.version>2.40.0</htmlunit.version>
         <mockito.version>3.3.3</mockito.version>
         <jna.version>5.3.1</jna.version><!-- should satisfy both testcontainers and mongodb -->
         <antlr.version>4.7.2</antlr.version>

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java
@@ -10,6 +10,7 @@ import java.util.Optional;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.regex.Pattern;
 
 import javax.crypto.SecretKey;
 import javax.crypto.spec.SecretKeySpec;
@@ -54,7 +55,8 @@ public class CodeAuthenticationMechanism extends AbstractOidcAuthenticationMecha
     private static final String STATE_COOKIE_NAME = "q_auth";
     private static final String SESSION_COOKIE_NAME = "q_session";
     private static final String POST_LOGOUT_COOKIE_NAME = "q_post_logout";
-    private static final String COOKIE_DELIM = "___";
+    private static final String COOKIE_DELIM = "|";
+    private static final Pattern COOKIE_PATTERN = Pattern.compile("\\" + COOKIE_DELIM);
 
     private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity securityIdentity,
             String accessToken,
@@ -86,13 +88,13 @@ public class CodeAuthenticationMechanism extends AbstractOidcAuthenticationMecha
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
-            String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
+            String[] tokens = COOKIE_PATTERN.split(sessionCookie.getValue());
             String idToken = tokens[0];
             String accessToken = tokens[1];
             String refreshToken = tokens[2];
 
             TenantConfigContext configContext = resolver.resolve(context, true);
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
                     .map(new Function<SecurityIdentity, SecurityIdentity>() {
                         @Override
                         public SecurityIdentity apply(SecurityIdentity identity) {
@@ -111,23 +113,21 @@ public class CodeAuthenticationMechanism extends AbstractOidcAuthenticationMecha
 
                             Throwable cause = throwable.getCause();
 
-                            // we should have proper exception hierarchy to represent token expiration errors
-                            if (cause != null && !cause.getMessage().equalsIgnoreCase("expired token")) {
-                                throw new AuthenticationCompletionException(throwable);
+                            if (cause != null && !"expired token".equalsIgnoreCase(cause.getMessage())) {
+                                LOG.debugf("Authentication failure: %s", cause);
+                                throw new AuthenticationCompletionException(cause);
                             }
-
-                            // try silent refresh if required
-                            SecurityIdentity identity = null;
-
-                            if (configContext.oidcConfig.token.refreshExpired) {
-                                identity = trySilentRefresh(configContext, idToken, refreshToken, context,
-                                        identityProviderManager);
+                            if (!configContext.oidcConfig.token.refreshExpired) {
+                                LOG.debug("Token has expired, token refresh is not allowed");
+                                throw new AuthenticationCompletionException(cause);
                             }
-
+                            LOG.debug("Token has expired, trying to refresh it");
+                            SecurityIdentity identity = trySilentRefresh(configContext, idToken, refreshToken, context,
+                                    identityProviderManager);
                             if (identity == null) {
-                                throw new AuthenticationFailedException(throwable);
+                                LOG.debug("SecurityIdentity is null after a token refresh");
+                                throw new AuthenticationCompletionException();
                             }
-
                             return identity;
                         }
                     });
@@ -196,7 +196,7 @@ public class CodeAuthenticationMechanism extends AbstractOidcAuthenticationMecha
                 return Uni.createFrom().failure(new AuthenticationCompletionException());
             } else if (context.queryParam("pathChecked").isEmpty()) {
                 // This is an original redirect from IDP, check if the request path needs to be updated
-                String[] pair = stateCookie.getValue().split(COOKIE_DELIM);
+                String[] pair = COOKIE_PATTERN.split(stateCookie.getValue());
                 if (pair.length == 2) {
                     // The extra path that needs to be added to the current request path
                     String extraPath = pair[1];

--- a/integration-tests/oidc-code-flow/src/main/resources/application.properties
+++ b/integration-tests/oidc-code-flow/src/main/resources/application.properties
@@ -14,6 +14,7 @@ quarkus.oidc.authentication.cookie-path=/
 quarkus.oidc.authentication.extra-params.max-age=60
 quarkus.oidc.application-type=web-app
 
+
 # Tenant which does not need to restore a request path after redirect, client_secret_post method
 quarkus.oidc.tenant-1.auth-server-url=${keycloak.url}/realms/quarkus
 quarkus.oidc.tenant-1.client-id=quarkus-app
@@ -98,4 +99,5 @@ quarkus.http.auth.permission.post-logout.policy=permit
 quarkus.http.cors=true
 
 quarkus.log.category."io.quarkus.oidc.runtime.CodeAuthenticationMechanism".level=TRACE
+quarkus.log.category."io.vertx.ext.auth.oauth2.impl.OAuth2UserImpl".level=TRACE
 #quarkus.log.category."org.apache.http".level=TRACE

--- a/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
+++ b/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
@@ -14,7 +14,6 @@ import java.util.concurrent.Callable;
 import java.util.concurrent.TimeUnit;
 
 import org.hamcrest.Matchers;
-import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
 import com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException;
@@ -139,7 +138,6 @@ public class CodeFlowTest {
     }
 
     @Test
-    @Disabled
     public void testRPInitiatedLogout() throws IOException {
         try (final WebClient webClient = createWebClient()) {
             HtmlPage page = webClient.getPage("http://localhost:8081/tenant-logout");
@@ -415,7 +413,6 @@ public class CodeFlowTest {
     }
 
     @Test
-    @Disabled
     public void testAccessTokenInjection() throws IOException {
         try (final WebClient webClient = createWebClient()) {
             HtmlPage page = webClient.getPage("http://localhost:8081/index.html");
@@ -502,11 +499,11 @@ public class CodeFlowTest {
     }
 
     private String getStateCookieStateParam(WebClient webClient, String tenantId) {
-        return getStateCookie(webClient, tenantId).getValue().split("___")[0];
+        return getStateCookie(webClient, tenantId).getValue().split("\\|")[0];
     }
 
     private String getStateCookieSavedPath(WebClient webClient, String tenantId) {
-        String[] parts = getStateCookie(webClient, tenantId).getValue().split("___");
+        String[] parts = getStateCookie(webClient, tenantId).getValue().split("\\|");
         return parts.length == 2 ? parts[1] : null;
     }
 
@@ -515,6 +512,6 @@ public class CodeFlowTest {
     }
 
     private String getIdToken(Cookie sessionCookie) {
-        return sessionCookie.getValue().split("___")[0];
+        return sessionCookie.getValue().split("\\|")[0];
     }
 }

--- a/integration-tests/oidc-tenancy/src/test/java/io/quarkus/it/keycloak/BearerTokenAuthorizationTest.java
+++ b/integration-tests/oidc-tenancy/src/test/java/io/quarkus/it/keycloak/BearerTokenAuthorizationTest.java
@@ -153,7 +153,7 @@ public class BearerTokenAuthorizationTest {
     }
 
     private String getStateCookieSavedPath(WebClient webClient, String tenantId) {
-        String[] parts = getStateCookie(webClient, tenantId).getValue().split("___");
+        String[] parts = getStateCookie(webClient, tenantId).getValue().split("\\|");
         return parts.length == 2 ? parts[1] : null;
     }
 }

